ghidra/p_code_extractor/PcodeExtractor.java

```diff
@@ -0,0 +1,1091 @@
+            return new Term<Jmp>(jmpTid, new Jmp(ExecutionType.JmpType.GOTO, mnemonic, createLabel(mnemonic, pcodeOp, null), pCodeCount));
+        } else if (mnemonic.equals("RETURN")) {
+            return new Term<Jmp>(jmpTid, new Jmp(ExecutionType.JmpType.RETURN, mnemonic, createLabel(mnemonic, pcodeOp, null), pCodeCount));
+        }
+
+        Term<Jmp> call = new Term<Jmp>(jmpTid, new Jmp(ExecutionType.JmpType.CALL, mnemonic, createCall(instr, mnemonic, pcodeOp), pCodeCount));
+        call = checkIfCallindResolved(call);
+
+        return call;
+    }
+
+
+    /**
+     * 
+     * @param call: indirect call
+     * @return: direkt call or indirekt call
+     * 
+     * Checks whether the indirect call could have been resolved and casts it into a direct call
+     */
+    protected Term<Jmp> checkIfCallindResolved(Term<Jmp> call) {
+        if (call.getTerm().getMnemonic().equals("CALLIND")) {
+            if (call.getTerm().getCall().getTarget().getIndirect() == null) {
+                call.getTerm().setMnemonic("CALL");
+            }
+        }
+
+        return call;
+    }
+
+
+    /**
+     * @param pCodeCount: Pcode index in current block
+     * @param pcodeOp:    Pcode instruction
+     * @param instrAddr:  Assembly instruction address
+     * @return: new Def Term
+     * 
+     * Creates a Def Term with an unique TID consisting of the prefix def, its instruction address and the index of the pcode in the block.
+     */
+    protected Term<Def> createDefTerm(int pcodeIndex, PcodeOp pcodeOp, Address instrAddr) {
+        Tid defTid = new Tid(String.format("instr_%s_%s", instrAddr.toString(), pcodeIndex), instrAddr.toString());
+        if (pcodeOp.getMnemonic().equals("STORE")) {
+            return new Term<Def>(defTid, new Def(createExpression(pcodeOp), pcodeIndex));
+            // cast copy instructions that have address outputs into store instructions
+        } else if (pcodeOp.getMnemonic().equals("COPY") && pcodeOp.getOutput().isAddress()) {
+            return new Term<Def>(defTid, new Def(new Expression("STORE", null, createVariable(pcodeOp.getOutput(), false), createVariable(pcodeOp.getInput(0), false)), pcodeIndex));
+        }
+        return new Term<Def>(defTid, new Def(createVariable(pcodeOp.getOutput(), false), createExpression(pcodeOp), pcodeIndex));
+    }
+
+
+    /**
+     * @param node: Varnode source for Variable
+     * @return: new Variable
+     * 
+     * Set register name based on being a register, virtual register, constant or ram address.
+     * In case it is a virtual register, prefix the name with $U.
+     * In case it is a constant, remove the const prefix from the constant.
+     */
+    protected Variable createVariable(Varnode node, Boolean isArgument) {
+        Variable var = new Variable();
+        if (node.isRegister()) {
+            String mnemonic = getRegisterMnemonic(node);
+            if(isArgument && mnemonic.equals("AL")){
```

Comment:
We cannot check for specific register names here, since this is not architecture-independent! Also, I would move the handling of this special case into its own function that only gets called when handling external symbols.

Reply:
Done